File: top_readers.py
"""Summer reading program: find the top five readers by points.

Scoring: 50 points per book read + 1 point per page.
Data file: https://public.karat.io/content/test/test_file.txt
Each line: <reader id>,<reader name>,<book name>,<number of pages>
"""

import csv
import sys
from collections import defaultdict
from urllib.request import urlopen

DATA_URL = "https://public.karat.io/content/test/test_file.txt"
POINTS_PER_BOOK = 50


def read_lines(source):
    """Yield lines from a local path or, by default, the remote data file."""
    if source:
        with open(source, newline="", encoding="utf-8") as fh:
            yield from fh
    else:
        with urlopen(DATA_URL) as resp:
            for raw in resp:
                yield raw.decode("utf-8")


def score_readers(lines):
    """Return {reader_id: total_points} from CSV lines.

    Book names may contain commas, so we treat the last field as pages and the
    first as the id, joining anything in between as the book name.
    """
    points = defaultdict(int)
    for row in csv.reader(lines):
        if len(row) < 4:
            continue
        reader_id, pages = row[0].strip(), row[-1].strip()
        if not pages.isdigit():
            continue  # skip header or malformed rows
        points[reader_id] += POINTS_PER_BOOK + int(pages)
    return points


def top_readers(points, n=5):
    # Sort by points descending; ties broken by id for stable output.
    return sorted(points.items(), key=lambda kv: (-kv[1], kv[0]))[:n]


def main():
    source = sys.argv[1] if len(sys.argv) > 1 else None
    points = score_readers(read_lines(source))
    for reader_id, total in top_readers(points):
        print(reader_id, total)


if __name__ == "__main__":
    main()

File: top_readers.ts
/**
 * Summer reading program: find the top five readers by points.
 *
 * Scoring: 50 points per book read + 1 point per page.
 * Data file: https://public.karat.io/content/test/test_file.txt
 * Each line: <reader id>,<reader name>,<book name>,<number of pages>
 *
 * Run: npx tsx top_readers.ts            (fetches from the URL)
 *      npx tsx top_readers.ts file.txt   (uses a local file)
 */

import { readFile } from "node:fs/promises";

const DATA_URL = "https://public.karat.io/content/test/test_file.txt";
const POINTS_PER_BOOK = 50;

async function readContent(source?: string): Promise<string> {
  const target = source ?? DATA_URL;
  // Fetch over the network for URLs; otherwise read a local file.
  if (/^https?:\/\//i.test(target)) {
    const resp = await fetch(target);
    if (!resp.ok) throw new Error(`Failed to fetch data: ${resp.status}`);
    return resp.text();
  }
  return readFile(target, "utf-8");
}

function scoreReaders(content: string): Map<string, number> {
  const points = new Map<string, number>();
  for (const line of content.split(/\r?\n/)) {
    if (!line.trim()) continue;
    // Book names may contain commas: id is the first field, pages the last.
    const fields = line.split(",");
    if (fields.length < 4) continue;
    const readerId = fields[0].trim();
    const pages = Number(fields[fields.length - 1].trim());
    if (!Number.isInteger(pages)) continue; // skip header / malformed rows
    points.set(readerId, (points.get(readerId) ?? 0) + POINTS_PER_BOOK + pages);
  }
  return points;
}

function topReaders(
  points: Map<string, number>,
  n = 5,
): Array<[string, number]> {
  // Sort by points descending; ties broken by id for stable output.
  return [...points.entries()]
    .sort((a, b) => b[1] - a[1] || a[0].localeCompare(b[0]))
    .slice(0, n);
}

async function main(): Promise<void> {
  const source = process.argv[2];
  const points = scoreReaders(await readContent(source));
  for (const [readerId, total] of topReaders(points)) {
    console.log(readerId, total);
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
